04_Scheduling/04_Locks.md

```diff
@@ -0,0 +1,201 @@
+
+Imagine we have a shared resource that can be accessed at a specific address. This resource could be anything from MMIO, a buffer or some variable, the important part is that multiple threads *can* access it at the same time. 
+
+For our example we're going to say this resouce is a NS16550 uart at address `0xDEADBEEF`. If you're not familiar with this type of uart device, it's the de facto standard for serial devices. The COM ports on x86 use one of these, as do many other platforms. 
+
+The key things to know are that if you write a byte at that address, it will be sent over the serial port to whatever is one the other send. So if you want to send a message, you must send it one character at a time, at the address specified (`0xDEADBEEF`).
+
+## The Problem
+
+Let's say we use this serial port to for log messages, with a function like the following:
+
+```c
+void serial_log(const char* msg) {
+    volatile char* resource = (char*)0xDEADBEEF;
+    for (size_t i = 0; msg[i] != 0; i++)
+        *resource = msg[i];
+}
+```
+
+Note that in reality you should check that the uart is ready to receive the next byte, and there is some setup you should do before sending. You will need a little more code than this for a complete uart driver, but that's outside the scope of this chapter.
+
+Now let's say we have a thread that wants to log a message:
+
+```c
+void thead_one() {
+    serial_log("I am the first string");
+}
+```
+
+This would work as expected. We would see `I am the first string` on the serial output.
+
+Now lets introduce a second thread, that does something similar:
+
+```c
+void thead_two() {
+    serial_log("while I am the second string");
+}
+```
+
+What would we expect to see on the serial output? We dont know! It's essentially non-deterministic, since we can't know how these will be scheduled. Each thread may get to write the full string before the other is scheduled, but more like they will get in the way of each other.
+
+![Tasks execution sequence](Images/taskssequence.png)
+
+One example of what we could see is `Iwh aI ammi  lethe secfionrsd t stristngring`. This contains all the right characters but it's completely unreadable. The image below shows what could happen:
+
+![Shared Resource Sequence](Images/sharedressequence.png)
+
+What we'd expect to see is one of two outcomes: `I am the first string while I am the second` or `while I am the second I am the first string`.
+
+This situation we've described is an example of a _race condition_. The order of accesses the the shared resource (the uart) matters here, so we need a way to protect it. This where a lock comes in.
+
+## Implementing A Lock
+
+A _lock_ provides us with something called mutual exclusion: only one thread can hold the lock at a time, while the rest must wait for the lock to be free again before they can take. While a thread is holding the lock, it's allowed to access the shared resource.
```

Comment:
not sure if something is missing at the end of the first sentence: 

while the rest must wait for the lock to be free again before they can take...